### ISSUE TITLE
[no-unnecessary-type-arguments] False positive when function parameter type takes precedence over default generic type

### ISSUE THREAD

**Issue by jpribyl** [typescript-eslint/typescript-eslint#3479]
**Repro**
See [playground](https://www.typescriptlang.org/play?ts=4.2.3#code/CYUwxgNghgTiAEAzArgOzAFwJYHtVIB4AVeAXngGcMYtUBzAPgAooAueIgSnalQE8A3ACghiJgEZOA+AHoZHeLUQgYcYPCgV4qZAFsARivgALFSAA02nPAx8ADiApgadjPBUwcMUQSo16zJKy8iQUDmBYiFgg6pqU1LR0JmaWtg5OLm4eXvAA1FYAtGio4I4UsHwFaSAFsHR6IKgYWohQEBQIdjgUWNgAbiBAA)

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-type-arguments": "error",
  }
}
```

```typescript
declare function f<T = string>(a: T): any;

f(1); // T inferred as number here, no typescript error
f<string>(1) // T specified as string here, typescript error + no-unnecessary-type-arguments false positive
```

**Expected Result**
No errors, TS does not compile `f<string>` in the same way that it compiles `f`

**Actual Result**
Error on line 4:
```TS
f<string>(1) // @typescript-eslint/no-unnecessary-type-arguments: This is the default value for this type parameter, so it can be omitted.
```

**Additional Info**
debug output is extremely long and doesn't seem super helpful here

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.22.1` |
| `@typescript-eslint/parser`        | `4.22.1` |
| `TypeScript`                       | `4.2.4` |
| `ESLint`                           | `7.25.0` |
| `node`                             | `14.17.0` |

**Comment by bradzacher** [comment]
The code you provided is not valid TS code - so we don't particularly care if the rule errors on it.

Please provide a more concrete and compile-time valid example of code that uses this pattern.

**Comment by jpribyl** [comment]
@bradzacher hmm I see, so you're saying that because it will only ever appear in conjunction with a TS error then it doesn't matter that the rule errors incorrectly?

The use case here was designing typed guard-rails around an exposed API for a package with nested functions.

I'll try to think of a simple way to recreate this for you

**Comment by bradzacher** [comment]
Correct.
If typescript is erroring then your code is invalid so your build is blocked.
At that point you are in the realm of undefined behaviour with most lint rules.

As another example - often times when a type error occurs TS will go to `any` - which would trigger our `no-unsafe-*` lint rules. But we'll never add handling for that case.
